### ISSUE TITLE
Line dashes on Windows 200% display scale are half-size

### ISSUE THREAD

**Issue by Phillipus** [eclipse/gef-classic#206]
This affects Windows only, the problem is not seen on Mac or Linux.

I've not noticed this before now because I only recently purchased a hi-res monitor. I have it set to 200% display scaling.

In Archi, our app using GEF and Draw2d, we use connection lines that have dashes. Here's an example on Windows at 100% display scaling:

![a100](https://github.com/eclipse/gef-classic/assets/600504/f9656cfe-d50f-4501-bd27-81347ffa5264)

And here it is at 200% scaling:

![a200](https://github.com/eclipse/gef-classic/assets/600504/65eb14ff-4b01-4857-adbf-ee08cd2bd58a)

Notice how the dots are closer together in the 200% version.

This does not happen on Linux and Mac at 200% scaling, on those platforms the line dashes look the same as in the first screenshot.

**Comment by Phillipus** [comment]
Here's a Draw2d snippet to reproduce the problem.

Requirements:

- Windows
- Hi-res monitor capable of displaying at 100% and 200% scaling to compare.

```java
import org.eclipse.draw2d.LightweightSystem;
import org.eclipse.draw2d.Polyline;
import org.eclipse.draw2d.RectangleFigure;
import org.eclipse.draw2d.geometry.Point;
import org.eclipse.draw2d.geometry.Rectangle;
import org.eclipse.swt.SWT;
import org.eclipse.swt.graphics.Color;
import org.eclipse.swt.layout.FillLayout;
import org.eclipse.swt.widgets.Canvas;
import org.eclipse.swt.widgets.Display;
import org.eclipse.swt.widgets.Shell;

public class LineDashesDraw2d {

    public static void main(String[] args) {
        Display display = new Display();
        Shell shell = new Shell(display);
        shell.setLayout(new FillLayout());
        shell.setSize(300, 300);

        Canvas fc = new Canvas(shell, SWT.NONE);
        LightweightSystem lws = new LightweightSystem(fc);
        
        RectangleFigure root = new RectangleFigure();
        root.setBackgroundColor(new Color(null, 255, 255, 255));
        lws.setContents(root);
        
        RectangleFigure rect = new RectangleFigure();
        rect.setBackgroundColor(new Color(null, 255, 255, 255));
        rect.setBounds(new Rectangle(10, 10, 250, 150));
        rect.setLineWidth(1);
        rect.setLineStyle(SWT.LINE_CUSTOM);
        rect.setLineDash(new float[] { 4 });
        root.add(rect);

        Polyline line = new Polyline();
        line.addPoint(new Point(10, 85));
        line.addPoint(new Point(260, 85));
        line.setLineStyle(SWT.LINE_CUSTOM);
        line.setLineDash(new float[] { 4 });
        root.add(line);

        shell.open();
        while(!shell.isDisposed()) {
            if(!display.readAndDispatch()) {
                display.sleep();
            }
        }
    }
}
```

Windows 100% scaling:

![100](https://github.com/eclipse/gef-classic/assets/600504/e61b5fb1-14ab-4383-8124-df1bdceb1402)

Windows 200% scaling:

![200](https://github.com/eclipse/gef-classic/assets/600504/4d5c4d96-3efa-44f3-a1c3-9da4d959d6a8)

**Comment by Phillipus** [comment]
The underlying issue comes from SWT:

https://github.com/eclipse-platform/eclipse.platform.swt/issues/687

**Comment by azoitl** [comment]
Have you tried with intermediate scaling factors (e.g., 125% or 150%) also? Is this related to the problem I reported here: https://github.com/eclipse/gef-classic/issues/106#issuecomment-1433699170 ?

**Comment by Phillipus** [comment]
> Have you tried with intermediate scaling factors (e.g., 125% or 150%) also? Is this related to the problem I reported here: [#106 (comment)](https://github.com/eclipse/gef-classic/issues/106#issuecomment-1433699170) ?

Not yet. However, I did try running 4Diac IDE on Windows at 200% scaling to compare and the rendering of figures is broken (contents are displaced.)

**Comment by Phillipus** [comment]
This is fundamentally a Windows SWT issue, and I'm not sure whether it's a bug or a "feature".

For Archi, and my Draw2D fork, I've devised an ugly hack in `SWTGraphics`:

```java
public void setLineDash(float[] value, boolean adjustForWindowsHiDPI) {
    if (value != null) {
        // validate dash values
        for (int i = 0; i < value.length; i++) {
            if (value[i] <= 0) {
                SWT.error(SWT.ERROR_INVALID_ARGUMENT);
            }
        }
        
        if(adjustForWindowsHiDPI && "win32".equals(SWT.getPlatform()) &&
                org.eclipse.swt.internal.DPIUtil.getDeviceZoom() == 200) {
            for(int i = 0; i < value.length; i++) {
                value[i] *= 2;
            }
        }

        currentState.lineAttributes.dash = value.clone();
        currentState.lineAttributes.style = SWT.LINE_CUSTOM;
    } else {
        currentState.lineAttributes.dash = null;
        currentState.lineAttributes.style = SWT.LINE_SOLID;
    }
}
```

```java
public void setLineAttributes(LineAttributes lineAttributes, boolean adjustForWindowsHiDPI) {
    copyLineAttributes(currentState.lineAttributes, lineAttributes);
    
    // HACK HERE
    if(adjustForWindowsHiDPI && currentState.lineAttributes.dash != null &&
            "win32".equals(SWT.getPlatform()) && org.eclipse.swt.internal.DPIUtil.getDeviceZoom() == 200) {
        for(int i = 0; i < currentState.lineAttributes.dash.length; i++) {
            currentState.lineAttributes.dash[i] *= 2;
        }
    }
}
```

**Comment by azoitl** [comment]
> > Have you tried with intermediate scaling factors (e.g., 125% or 150%) also? Is this related to the problem I reported here: [#106 (comment)](https://github.com/eclipse/gef-classic/issues/106#issuecomment-1433699170) ?
> 
> Not yet. However, I did try running 4Diac IDE on Windows at 200% scaling to compare and the rendering of figures is broken (contents are displaced.)

I know that. But AFAIK this only happens if you have two screens with different scaling settings. But I don't have a clue where it is coming from. Especially as only parts are off.

**Comment by azoitl** [comment]
@Phillipus is this issue also improved throw your fix in PR #248?

**Comment by Phillipus** [comment]
This is now fixed upstream in SWT:

https://github.com/eclipse-platform/eclipse.platform.swt/commit/e83f6dd27049c44f06f9894d82253f857a89a9ba
